singer_sdk/helpers/_typing.py

```diff
@@ -275,46 +289,101 @@ def conform_record_data_types(  # noqa: C901
+
+    return rec
+
+
+def _conform_record_data_types(
+    input_object: Dict[str, Any], schema: dict, parent: Optional[str]
+) -> Tuple[Dict[str, Any], List[str]]:  # noqa: C901
+    """Translate values in record dictionary to singer-compatible data types.
+
+    Any property names not found in the schema catalog will be removed, and a single
+    warning will be logged listing each unmapped property name.
+
+    This is called recursively to process nested objects and arrays.
+
+    Args:
+        input_object: A single record
+        schema: JSON schema the given input_object is expected to meet
+        parent: '.' seperated path to this element from the object root (for logging)
+    """
+    output_object: Dict[str, Any] = {}
     unmapped_properties: List[str] = []
-    for property_name, elem in record.items():
+    for property_name, elem in input_object.items():
+        property_path = (
+            property_name if parent is None else parent + "." + property_name
+        )
         if property_name not in schema["properties"]:
-            unmapped_properties.append(property_name)
+            unmapped_properties.append(property_path)
             continue
 
         property_schema = schema["properties"][property_name]
-        if isinstance(elem, (datetime.datetime, pendulum.DateTime)):
-            rec[property_name] = to_json_compatible(elem)
-        elif isinstance(elem, datetime.date):
-            rec[property_name] = elem.isoformat() + "T00:00:00+00:00"
-        elif isinstance(elem, datetime.timedelta):
-            epoch = datetime.datetime.utcfromtimestamp(0)
-            timedelta_from_epoch = epoch + elem
-            rec[property_name] = timedelta_from_epoch.isoformat() + "+00:00"
-        elif isinstance(elem, datetime.time):
-            rec[property_name] = str(elem)
-        elif isinstance(elem, bytes):
-            # for BIT value, treat 0 as False and anything else as True
-            bit_representation: bool
-            if is_boolean_type(property_schema):
-                bit_representation = elem != b"\x00"
-                rec[property_name] = bit_representation
-            else:
-                rec[property_name] = elem.hex()
-        elif is_boolean_type(property_schema):
-            boolean_representation: Optional[bool]
-            if elem is None:
-                boolean_representation = None
-            elif elem == 0:
-                boolean_representation = False
-            else:
-                boolean_representation = True
-            rec[property_name] = boolean_representation
+        if isinstance(elem, list) and is_uniform_list(property_schema):
+            item_schema = property_schema["items"]
+            output = []
+            for item in elem:
+                if is_object_type(item_schema) and isinstance(item, dict):
+                    output_item, sub_unmapped_properties = _conform_record_data_types(
```

Comment:
Do you think this recursiveness may cause performance issues for deeply nested records or large numbers of records?

If so, what would you say about limiting conforming via a parameter set by the tap developer, perhaps at the tap level?

For example:

```
class MyStream(Stream):
   conform_records_max_depth = 10
```

Reply:
Type conformance is probably the slowest part of the codebase, even for non-recursive schemas.
So if people were relying on the behaviour of it not conforming nested types for performance reasons, I suppose that could cause performance to go down in some cases.

Still, I don't think you would want to address that by tweaking the max depth, since you probably only want none, root or all.
Perhaps it should instead be an enum;
- Only conform root fields
- Conform all fields recursively
- Do not conform any fields
?

One of the reasons I want this merged to master was actaully because I'd like to start experimenting more in that area of the codebase to see if conformance performance can be improved 😁

Reply:
> Perhaps it should instead be an enum;
> 
> * Only conform root fields
> * Conform all fields recursively
> * Do not conform any fields
>   ?

@Jack-Burnett I like that. Go for it!

Reply:
Done!